lib/spack/spack/installer.py

```diff
@@ -0,0 +1,1554 @@
+            tty.die('\'{0}\' is not an allowed phase for package {1}'
+                    .format(self.pkg.last_phase, self.pkg.name))
+
+    def _cleanup_all_tasks(self):
+        """Cleanup all build tasks to include releasing their locks."""
+        for pkg_id in self.locks:
+            self._release_lock(pkg_id)
+
+        for pkg_id in self.failed:
+            self._cleanup_failed(pkg_id)
+
+        ids = list(self.build_tasks.keys())
+        for pkg_id in ids:
+            try:
+                self._remove_task(pkg_id)
+            except Exception:
+                pass
+
+    def _cleanup_failed(self, pkg_id):
+        """
+        Cleanup any failed mark for the package
+
+        Args:
+            pkg_id (str): identifier for the failed package
+        """
+        if pkg_id in self.failed:
+            lock = self.failed[pkg_id]
+            if lock is not None:
+                err = "{0} exception when removing failure mark for {1}: {2}"
+                msg = 'Removing failure mark on {0}'
+                try:
+                    tty.verbose(msg.format(pkg_id))
+                    lock.release_write()
+                except AssertionError:
+                    pass
+                except Exception as exc:
+                    tty.warn(err.format(exc.__class__.__name__, pkg_id,
+                                        str(exc)))
+
+    def _cleanup_task(self, pkg, remove_task):
+        """
+        Cleanup the build task for the spec
+
+        Args:
+            pkg (spack.package.PackageBase): the package being installed
+            remove_task (bool): ``True`` if the build task should be removed
+        """
+        if remove_task:
+            self._remove_task(pkg.unique_id)
+
+        # Ensure we have a read lock to prevent others from uninstalling the
+        # spec during our installation.
+        self._ensure_read_locked(pkg)
+
+    def _ensure_install_ready(self, pkg):
+        """
+        Ensure the package is ready to install locally, which includes
+        already locked.
+
+        Args:
+            pkg (spack.package.PackageBase): the package being locally
+                installed
+        """
+        pre = "{0} cannot be installed locally:".format(pkg.unique_id)
+
+        # External packages cannot be installed locally.
+        if pkg.spec.external:
+            # TODO: If keep, use a custom exception
+            raise spack.error.SpackError('{0} {1}'.format(pre, 'is external'))
+
+        # Upstream packages cannot be installed locally.
+        if pkg.installed_upstream:
+            # TODO: If keep, use a custom exception
+            raise spack.error.SpackError('{0} {1}'.format(pre, 'is upstream'))
+
+        # The package must have a prefix lock at this stage.
+        if pkg.unique_id not in self.locks:
+            # TODO: If keep, use a custom exception
+            raise spack.error.SpackError('{0} {1}'.format(pre, 'not locked'))
+
+    def _ensure_read_locked(self, pkg):
```

Comment:
I guess this is also a matter of style, but for me this would read better if the case of `nolock` or downgrade from `write` lock were clearly separated. Right now they are mixed in the same flow under conditionals and that necessitates to "context switch" frequently when reading the code.

Reply:
The early version of the two functions were so similar  it seemed "easier" to merge them into one operation.

Is there anything I could to in the comments to make this clearer?

Reply:
Nvm.  I refactored this and `_ensure_write_locked` into `_ensure_locked`.